### ISSUE TITLE
ty gives different result type for `divmod(Fraction(), 1)` than other checkers [invalid-return-type]

### ISSUE THREAD

**Issue by jepler** [astral-sh/ty#4285]
### Summary

The following code is clean of type errors under mypy --strict (2.3.1), pyrefly (1.1.1), and pyright (1.1.411) but gives an error in ty:
```py
from fractions import Fraction


def f(x: Fraction) -> int:
    whole, _= divmod(x, 1)
    return whole
```

The diagnostic from ty is:
```
error[invalid-return-type]: Return type does not match returned value
 --> divmod_type.py:6:12
  |
4 | def f(x: Fraction) -> int:
  |                       --- Expected `int` because of return type
5 |     whole, _= divmod(x, 1)
6 |     return whole
  |            ^^^^^ expected `int`, found `float`
info: element `float*` of union `int | float*` is not assignable to `int`

Found 1 diagnostic
```

This is the minimized case of a problem I encountered in real code when comparing ty & mypy.

### Version

ty 0.0.72

**Comment by carljm** [comment]
Thanks for the report!

Surprisingly (to me), this issue doesn't actually have anything to do with the int/float special case, it's purely a generic inference issue. It reproduces with this self-contained example:

```py
from typing import Protocol, overload

class Base: ...
class Derived(Base): ...

class SupportsOperation[T, R](Protocol):
    def operation(self, value: T, /) -> R: ...

class Operation:
    @overload
    def operation(self, value: Derived, /) -> Derived: ...
    @overload
    def operation(self, value: Base, /) -> Base: ...
    def operation(self, value: Base, /) -> Base:
        return value

def apply[T, R](value: SupportsOperation[T, R], argument: T) -> R:
    return value.operation(argument)

reveal_type(Operation().operation(Derived()))  # Derived
reveal_type(apply(Operation(), Derived()))      # Base; should be Derived
```

This is ultimately mostly #3557, and astral-sh/ruff#26712 does fix it. From the `apply(Operation(), Derived())` call we ultimately conclude that both `Derived` and `Base` are valid solutions for `T`. But we should then intersect those to infer the type as `Derived`, and instead we union them to infer the type as `Base`.

**Comment by jepler** [comment]
Thanks -- I was unable to understand the underlying cause well enough to find #3557, but it's good to know that it's on devs radar and that a fix is in the works.
